File: gamelog-frontend/src/services/gameService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api';

const gameService = {

  // [RAWG API] Add game to library by rawgAPI
  addToLibrary: async (rawgId) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/games/add-library/${rawgId}`);
      return response.data;
    } catch (error) {
      throw new Error('Failed to add game to library');
    }
  },

  // [RAWG API] Add game to wishlist by rawgAPI
  addToWishlist: async (rawgId) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/games/add-wishlist/${rawgId}`);
      return response.data;
    } catch (error) {
      throw new Error('Failed to add game to wishlist');
    }
  },

  // Get all library games count
  getAllLibraryGames: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/games/library`, {
        params: { page: 0, size: 1000 }
      });
      return response.data.content || [];
    } catch (error) {
      throw new Error('Failed to fetch all library games');
    }
  },

  // Get all library games
  getLibraryGames: async (page = 0, size = 8, status = 'ALL', searchTerm = '') => {
    try {
      const response = await axios.get(`${API_BASE_URL}/games/library`, {
        params: {
          page,
          size,
          status: status !== 'ALL' ? status : 'ALL',
          search: searchTerm
        }
      });
      return response.data;
    } catch (error) {
      throw new Error('Failed to fetch library games');
    }
  },

  // Get all wishlist games count
  getAllWishlistGames: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/games/wishlist`, {
        params: { page: 0, size: 1000 }
      });
      return response.data.content || [];
    } catch (error) {
      throw new Error('Failed to fetch all wishlist games');
    }
  },

  // Get all wishlist games
  getWishlistGames: async (page = 0, size = 8, search = '') => {
    const response = await axios.get(`${API_BASE_URL}/games/wishlist`, {
      params: { page, size, search }
    });
    return response.data;
  },

  // Update game
  updateGame: async (gameId, updateData) => {
    try {
      const requestBody = {
        playedOn: updateData.playedOn || null,
        status: updateData.status,
        rating: updateData.rating || null,
        notes: updateData.notes || null,
        completedAt: updateData.completedAt || null
      };

      const response = await axios.put(`${API_BASE_URL}/games/${gameId}`, requestBody);
      return response.data;
    } catch (error) {
      throw new Error('Failed to update game');
    }
  },

  // Delete game from Library
  deleteGame: async (gameId) => {
    try {
      await axios.delete(`${API_BASE_URL}/games/${gameId}`);
    } catch (error) {
      throw new Error('Failed to delete game');
    }
  },

  // [RAWG API] Search games
  searchGames: async (query) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/games/search?query=${encodeURIComponent(query)}`);
      return response.data;
    } catch (error) {
      throw new Error('Failed to search games');
    }
  }
};

export default gameService;

File: gamelog/src/main/java/com/matkon/gamelog/services/GameService.java
package com.matkon.gamelog.services;

import com.matkon.gamelog.data.game.Game;
import com.matkon.gamelog.data.game.GameReleaseFilter;
import com.matkon.gamelog.data.game.GameStatus;
import com.matkon.gamelog.data.game.dto.GameForWishlistDto;
import com.matkon.gamelog.data.game.dto.GameSaveResultDto;
import com.matkon.gamelog.data.game.dto.GameSearchResultDto;
import com.matkon.gamelog.data.game.dto.GameUpdateRequestDto;
import com.matkon.gamelog.data.sync.ChangeDetail;
import com.matkon.gamelog.data.sync.FieldChange;
import com.matkon.gamelog.data.sync.SyncResultDto;
import com.matkon.gamelog.data.sync.SyncUtils;
import com.matkon.gamelog.repos.GameRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class GameService {

    private final GameRepository gameRepository;
    private final RawgClientService rawgClientService;

    public GameService(GameRepository gameRepository, RawgClientService rawgClientService) {
        this.gameRepository = gameRepository;
        this.rawgClientService = rawgClientService;
    }

    public Page<Game> getWishlistGames(int page, int size, String searchTerm) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "updatedAt"));

        if (searchTerm != null && !searchTerm.isBlank()) {
            return gameRepository.findWishlistGames(GameStatus.WISHLIST, searchTerm, pageable);
        }

        return gameRepository.findWishlistGames(GameStatus.WISHLIST, searchTerm, pageable);
    }

    public Page<GameForWishlistDto> getWishlistGamesDashboard(int page, int size, String sort, GameReleaseFilter gameReleaseFilter) {
        String[] sortParts = sort.split(",");
        String field = sortParts[0].trim();
        Sort.Direction direction = (sortParts.length > 1 && "desc".equalsIgnoreCase(sortParts[1]))
                ? Sort.Direction.DESC
                : Sort.Direction.ASC;

        Pageable pageable = PageRequest.of(page, size, Sort.by(direction, field));
        LocalDate today = LocalDate.now();

        Page<Game> games = switch (gameReleaseFilter) {
            case RELEASED_ONLY ->
                    gameRepository.findByStatusAndReleaseDateLessThanEqual(GameStatus.WISHLIST, today, pageable);
            case NOT_RELEASED_ONLY ->
                    gameRepository.findByStatusAndReleaseDateAfter(GameStatus.WISHLIST, today, pageable);
            case TBA -> gameRepository.findByStatusAndReleaseDateIsNull(GameStatus.WISHLIST, pageable);
            default -> gameRepository.findByStatus(GameStatus.WISHLIST, pageable);
        };

        return games.map(GameForWishlistDto::fromEntity);
    }


    public Page<Game> getLibraryGames(int page, int size, String status, String searchTerm) {
        Pageable pageable = PageRequest.of(page, size);

        GameStatus dbStatus;
        if (status == null || "ALL".equals(status) || status.trim().isEmpty()) {
            dbStatus = null;
        } else {
            try {
                dbStatus = GameStatus.valueOf(status.toUpperCase());
            } catch (IllegalArgumentException e) {
                throw new IllegalArgumentException("Invalid status: " + status);
            }
        }

        String dbSearchTerm = (searchTerm == null || searchTerm.trim().isEmpty()) ? null : searchTerm;

        return gameRepository.findLibraryGames(dbStatus, dbSearchTerm, pageable);
    }

    public List<GameSearchResultDto> searchGames(String query) {
        return rawgClientService.searchGames(query);
    }

    public GameSaveResultDto saveGame(Long rawgId, GameStatus gameStatus) {
        Optional<Game> existingGame = gameRepository.findByRawgId(rawgId);
        if (existingGame.isPresent()) {
            return new GameSaveResultDto(
                    existingGame.get(),
                    true,
                    "Game already exists in the library"
            );
        }

        try {
            Game game = rawgClientService.getGameDetails(rawgId);

            if (game != null) {
                game.setStatus(gameStatus);
                Game savedGame = gameRepository.save(game);
                return new GameSaveResultDto(
                        savedGame,
                        false,
                        "Game added successfully"
                );
            }
        } catch (Exception e) {
            System.err.println("Error fetching game: " + e.getMessage());
            throw new RuntimeException("Error adding game to library");
        }

        throw new RuntimeException("Game not found with ID: " + rawgId);
    }

    public void deleteGame(Long gameId) {
        gameRepository.deleteById(gameId);
    }

    @Transactional
    public Game updateGame(Long id, GameUpdateRequestDto updateRequest) {
        Game existingGame = gameRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Game not found with id: " + id));

        existingGame.setPlatform(updateRequest.getPlatform());
        existingGame.setStatus(updateRequest.getStatus());
        existingGame.setRating(updateRequest.getRating());
        existingGame.setNotes(updateRequest.getNotes());
        existingGame.setCompletedAt(updateRequest.getCompletedAt());
        existingGame.setUpdatedAt(LocalDateTime.now());
        existingGame.setFavourite(updateRequest.isFavourite());

        return gameRepository.save(existingGame);
    }

    public SyncResultDto syncGamesByStatus(GameStatus gameStatus) {
        List<Game> games = gameRepository.findAll()
                .stream()
                .filter(game -> game.getStatus() == gameStatus)
                .toList();

        int updatedCount = 0;
        List<ChangeDetail> changes = new ArrayList<>();

        for (Game localGame : games) {
            Game latestData = rawgClientService.getGameDetails(localGame.getRawgId());
            if (latestData == null) continue;

            List<FieldChange> fieldChanges = new ArrayList<>();
            boolean changed = false;

            // Release date
            if (SyncUtils.areDatesDifferent(localGame.getReleaseDate(), latestData.getReleaseDate())) {
                fieldChanges.add(new FieldChange("Release_Date", String.valueOf(localGame.getReleaseDate()), String.valueOf(latestData.getReleaseDate())));
                localGame.setReleaseDate(latestData.getReleaseDate());
                changed = true;
            }

            // Title
            if (SyncUtils.areStringsDifferent(localGame.getTitle(), latestData.getTitle())) {
                fieldChanges.add(new FieldChange("Title", localGame.getTitle(), latestData.getTitle()));
                localGame.setTitle(latestData.getTitle());
                changed = true;
            }

            // Image url
            if (SyncUtils.areStringsDifferent(localGame.getImageUrl(), latestData.getImageUrl())) {
                fieldChanges.add(new FieldChange("Image_Url", localGame.getImageUrl(), latestData.getImageUrl()));
                localGame.setImageUrl(latestData.getImageUrl());
                changed = true;
            }

            if (changed) {
                gameRepository.save(localGame);
                updatedCount++;
                changes.add(new ChangeDetail(localGame.getId(), localGame.getTitle(), fieldChanges));
            }
        }

        return new SyncResultDto(games.size(), updatedCount, changes);
    }
}